runtime/src/accounts_db.rs

```diff
@@ -4357,11 +4357,63 @@ impl AccountsDb {
                 for slot in start..end {
                     let sub_storages = snapshot_storages.get(slot);
                     if let Some(sub_storages) = sub_storages {
-                        for storage in sub_storages {
-                            let accounts = storage.accounts.accounts(0);
-                            accounts.into_iter().for_each(|stored_account| {
-                                scan_func(LoadedAccount::Stored(stored_account), &mut retval, slot)
-                            });
+                        let mut len = sub_storages.len();
+                        if len == 1 {
+                            for storage in sub_storages {
+                                let accounts = storage.accounts.accounts(0);
+                                accounts.into_iter().for_each(|stored_account| {
+                                    scan_func(
+                                        LoadedAccount::Stored(stored_account),
+                                        &mut retval,
+                                        slot,
+                                    )
+                                });
+                            }
+                        } else {
+                            // we have to call the scan_func in order of write_version within a slot if there are multiple storages per slot
```

Comment:
this messy code is required if we have > 1 append vec per slot. We need to write_version to be in increasing order in the final accumulated results.